### DISCUSSION TITLE
Why is migration publishing required?

### DISCUSSION THREAD

**Discussion by bilogic** [spatie/laravel-medialibrary#2594] [Q&A]
Just curious about why we need to publish the migration before running `php artisan migrate`.

As far as I can tell, the `media` table is not optional, why not auto discover it using `->loadMigrationsFrom(...)`.

Thank you.

**Comment by freekmurze** [comment]
We think the `loadMigrationsFrom` might be confusing for some as the migrations are not in your local project.
In generally we like all of the migrations to be in the `database/migrations` directory.

**Reply by bilogic** [reply]
I see, ok. 

I know it is on me, but sometimes I just cannot remember if I already ran `vendor:publish` or not. That leads to multiple `media` migrations being created which leads to exceptions when migrating.

Such a step also makes it harder for other packages to be built on top this package. i.e. adding our package to Laravel requires going through docs of the dependent packages. And these migrations are also not removed cleanly via `composer remove`

If we refer to https://github.com/eminiarts/nova-permissions, they actually copied https://github.com/spatie/laravel-permission migration into their own `database/migrations`, breaking the "inheritance", see https://github.com/eminiarts/nova-permissions/blob/master/database/migrations/create_permission_tables.php.stub

I don't know of any easy way around these challenges I faced. Also looked at Laravel's packages, they typically follow these steps:

```BASH
# composer require vendor/package
# php artisan migrate
# php artisan package:install // not all packages have this
# php artisan vendor:publish <for this package> // usually optional
```

Just a hope that if all packages have the same steps, it would lessen the need to constantly refer to docs. :)
Anyway, thanks for the wonderful packages from Spatie!

Thank you.
